src/dsc.h

```diff
@@ -417,46 +441,38 @@ class IdList {
         AllocForOneMore();
 
         // Copy-construct at the end of the list.
+        indexes[n] = n;
         new(&elem[n]) T(*t);
         ++n;
 
         return t->h;
     }
 
-    T * LowerBound(T const& t) {
-        if(IsEmpty()) {
-            return nullptr;
-        }
-        auto it = std::lower_bound(begin(), end(), t, Compare());
-        return it;
+    Iterator LowerBound(T const& t) {
+        return std::lower_bound(begin(), end(), t, Compare());
     }
 
-    T * LowerBound(H const& h) {
-        if(IsEmpty()) {
-            return nullptr;
-        }
-        auto it = std::lower_bound(begin(), end(), h, Compare());
-        return it;
+    Iterator LowerBound(H const& h) {
+        return std::lower_bound(begin(), end(), h, Compare());
```

Comment:
I wouldn't expect a meaningful difference, it boils down to a few pointer operations. Didn't you have to add a forward declaration and complicate `CompareId` because of that?

Reply:
Yes. I called `CompareId" the predicate.